mbq/atomiq/producers.py

```diff
@@ -49,8 +49,36 @@ def _is_running_within_transaction(self):
         save points and expect there at least to be 2: 1 from TestCase and 1 user-defined.
         """
         if RUNNING_TESTS:
-            if len(db_connection.savepoint_ids) < 2:
-                return False
+            import inspect
+            from django.test import TestCase
+            inSetup = False
+            inTestCase = False
+
+            # This loops through the call stack and sets "isSetUp" and "isTestCase"
+            for stack_frame in inspect.stack():
+                frame = stack_frame[0]
+                localVars = frame.f_locals
+                selfVar = localVars.get('self')
+                if selfVar:
+                    selfClass = getattr(selfVar, '__class__')
+                    if inspect.isclass(selfClass) and issubclass(selfClass, TestCase):
+                        inTestCase = True
+
+                clsClass = localVars.get('cls')
+                if inspect.isclass(clsClass) and issubclass(clsClass, TestCase):
+                    inTestCase = True
+
+                if stack_frame[3] in ['setUpTestData', 'setUp', 'setUpClass']:
+                    inSetup = True
+
+            if inTestCase:
+                if inSetup:
+                    # Don't enforce transactions in TestCase setup functions
+                    return True
+
+                elif len(db_connection.savepoint_ids) < 2:
+                    # Enforce 1 user transaction in TestCase unit tests
+                    return False
```

Comment:
This insane hack should probably be in its own function and have a ton of explanatory docstring, probably including an apology

Reply:
I mean this should be a really nice docstring

Reply:
Spend some time on it, see how nice you can make it

Reply:
yea let's take it out of producers and put it in utils for example

Reply:
I'm all for this 👍

Reply:
+1